module/network/RobotCommunication/src/RobotCommunication.cpp

```diff
@@ -0,0 +1,202 @@
+                            const std::vector<unsigned char>& payload = p.payload;
+                            RoboCup incoming_msg = NUClear::util::serialise::Serialise<RoboCup>::deserialise(payload);
+
+                            // filter out own messages
+                            if (global_config.player_id != incoming_msg.current_pose.player_id) {
+                                emit(std::make_unique<RoboCup>(std::move(incoming_msg)));
+                            }
+                        });
+                }
+            });
+
+        on<Every<2, Per<std::chrono::seconds>>,
+           Optional<With<Ball>>,
+           Optional<With<WalkState>>,
+           Optional<With<Kick>>,
+           Optional<With<Sensors>>,
+           Optional<With<Field>>,
+           Optional<With<GameState>>,
+           Optional<With<GlobalConfig>>>()
+            .then([this](const std::shared_ptr<const Ball>& loc_ball,
+                         const std::shared_ptr<const WalkState>& walk_state,
+                         const std::shared_ptr<const Kick>& kick,
+                         const std::shared_ptr<const Sensors>& sensors,
+                         const std::shared_ptr<const Field>& field,
+                         const std::shared_ptr<const GameState>& game_state,
+                         const std::shared_ptr<const GlobalConfig>& config) {
+                auto msg = std::make_unique<RoboCup>();
+
+                // Timestamp
+                msg->timestamp = NUClear::clock::now();
+
+                // State
+                int penalty_reason = game_state->data.self.penalty_reason;
+                switch (penalty_reason) {
+                    case 0: msg->state = 0; break;
+                    case 1: msg->state = 1; break;
+                    default: msg->state = 2; break;
+                }
+
+                // Current pose (Position, orientation, and covariance of the player on the field)
+                if (config) {
+                    msg->current_pose.player_id = config->player_id;
+
+                    if (sensors) {
+                        // Get our world transform
+                        Eigen::Isometry3d Htw(sensors->Htw);
+
+                        // If we have field information
+                        if (field) {
+                            // Transform the field state into Hfw
+                            Eigen::Isometry3d Hfw = Eigen::Isometry3d(field->Hfw);
+
+                            // Get our torso in field space
+                            Eigen::Isometry3d Hft = Hfw * Htw.inverse();
+                            Eigen::Vector3d rTFf  = Hft.translation();
+
+                            // Store our position from field to torso
+                            msg->current_pose.position.x = rTFf.x();
+                            msg->current_pose.position.y = rTFf.y();
+                            msg->current_pose.position.z = Hft.rotation().matrix().eulerAngles(0, 1, 2).z();
```

Comment:
I would suggest to use `MatrixToEulerIntrinsic` utility here instead of Eigen's `eulerAngles` 
```suggestion
                            msg->current_pose.position.z = MatrixToEulerIntrinsic(Hft.rotation()).z();
```

You will need to include `euler.hpp` utility for this.

Reply:
cheers